services/bots/github_apps.py

```diff
@@ -0,0 +1,242 @@
+import logging
+import random
+from datetime import datetime, timezone
+from typing import Dict, List, Optional
+
+from shared.django_apps.codecov_auth.models import Service
+from shared.github import is_installation_rate_limited
+from shared.typings.oauth_token_types import Token
+from shared.typings.torngit import GithubInstallationInfo
+
+from database.models import Owner, Repository
+from database.models.core import (
+    GITHUB_APP_INSTALLATION_DEFAULT_NAME,
+    GithubAppInstallation,
+)
+from helpers.exceptions import (
+    NoConfiguredAppsAvailable,
+)
+from services.bots.types import TokenWithOwner
+from services.github import get_github_integration_token
+from services.redis import get_redis_connection
+
+log = logging.getLogger(__name__)
+
+
+MAX_GITHUB_APP_SELECTION_WEIGHT = 1200
+
+
+def _get_installation_weight(installation: GithubAppInstallation) -> int:
+    """The weight for a given app installation.
+    Establishes an exponential ramp-up period for installations after being updated.
+    """
+    age = datetime.now(timezone.utc) - installation.created_at
+    if age.days >= 10:
+        return MAX_GITHUB_APP_SELECTION_WEIGHT
+    seconds_in_hour = 3600
+    age_hours = (age.seconds // seconds_in_hour) + age.days * 24
+    # Prevent clock differences from making the weight negative
+    return max(1, age_hours + 2**age.days)
+
+
+def _can_use_this_app(
+    app: GithubAppInstallation, installation_name: str, repository: Optional[Repository]
+) -> bool:
+    return (
+        app.name == installation_name
+        # We ignore apps that are not configured because those can't be used
+        and app.is_configured()
+        and (
+            # If there is a repo we want only the apps that cover said repo
+            (repository and app.is_repo_covered_by_integration(repository))
+            # If there is no repo we still need some true value
+            or (not repository)
+        )
+    )
+
+
+def _get_apps_from_weighted_selection(
+    owner: Owner, installation_name: str, repository: Optional[Repository]
+) -> List[GithubAppInstallation]:
+    """This function returns an ordered list of GithubAppInstallations that can be used to communicate with GitHub
+    in behalf of the owner. The list is ordered in such a way that the 1st element is the app to be used in Torngit,
+    and the subsequent apps are selected as fallbacks.
+
+    IF the repository is provided, the selected apps also cover the repo.
+    IF installation_name is not the default one, than the default codecov installation
+      is also selected as a possible fallback app.
+
+    Apps are selected randomly but assigned weights based on how recently they were created.
+    This means that older apps are selected more frequently as the main app than newer ones.
+    (up to 10 days, when the probability of being chosen is the same)
+    The random selection is done so we can distribute request load more evenly among apps.
```

Comment:
The feature request was to prevent normal operation from breaking if a new missconfigured app is introduced.

The idea is that a newly configured app could have been poorly configured. If selected it can derail normal operations. So by reducing the number of requests used by new installations we can maintain more regular operations (if it is missconfiguired we would see the errors appearing of course, but "most" requests would still be OK)

From that point on the ramp-up period is completely arbitrary.